File: Two Pointers/345. Reverse Vowels of a String.cpp
/*
345. Reverse Vowels of a String
Write a function that takes a string as input and reverse only the vowels of a string.

Example 1:
Given s = "hello", return "holle".

Example 2:
Given s = "leetcode", return "leotcede".

Note:
The vowels does not include the letter "y".

*/


class Solution {
public:
    string reverseVowels(string s) {
        unordered_set<char>dic = {'a','e','i','o','u','A','E','I','O','U'};
        int i = 0, j = s.size()-1;
        while(i<j){
            while(i<j && dic.find(s[i]) == dic.end()) i++;
            while(i<j && dic.find(s[j]) == dic.end()) j--;
            swap(s[i++],s[j--]);
        }
        return s;
    }
};


class Solution {
public:
    string reverseVowels(string s) {
        unordered_set<char>dic = {'a','e','i','o','u','A','E','I','O','U'};
        int i = 0, j = s.size()-1;
        while(i<j){
            while(i<j && dic.find(s[i]) == dic.end()) i++;
            while(i<j && dic.find(s[j]) == dic.end()) j--;
            if(i<j){
                s[i] ^= s[j];
                s[j] ^= s[i];
                s[i] ^= s[j];
                i++;j--;
            }
        }
        return s;
    }
};


class Solution {
public:
    string reverseVowels(string s) {
        int i = 0, j = s.size() - 1;
        while (i < j) {
            i = s.find_first_of("aeiouAEIOU", i);
            j = s.find_last_of("aeiouAEIOU", j);
            if (i < j) {
                swap(s[i++], s[j--]);
            }
        }
        return s;
    }
};

File: Two Pointers/345. Reverse Vowels of a String.py
"""
345. Reverse Vowels of a String

"""
class Solution(object):
    def reverseVowels(self, s):
        """
        :type s: str
        :rtype: str
        """
        v="AEIOUaeiou"
        i = 0
        j = len(s)-1
        li = list(s)
        while i<j:
            while i<j and li[i] not in v: i+=1
            while i<j and li[j] not in v: j-=1
            if i<j:
                li[i], li[j] = li[j], li[i]
            i+=1; j-=1
        
        return "".join(li)


class Solution(object):
    def reverseVowels(self, s):
        vowels = re.findall('(?i)[aeiou]', s)
        return re.sub('(?i)[aeiou]', lambda m: vowels.pop(), s)